src/network.py

```diff
@@ -120,25 +101,235 @@ def forward(self, enc_output):
+            stops.append(stop_pred)
+            # set stop_lens here!
+            outputs.append(dec_out)
+            input_ = outputs[-1]
+            i += 1
+
+            # double check this!
+            stop_mask = (stop_pred > .5).logical_and(stop_lens == 0)
+            stop_lens[stop_mask] = i
+            keep_gen = torch.all(stop_lens.neq(0)) and i < max_len
+
+        # Maybe this is a bit overkil...
+        stop_lens = stop_lens.masked_fill(stop_lens == 0, len(outputs))
+        pad_mask = sent_lens_to_mask(stop_lens, len(outputs))
+
+        res, res_stop = torch.stack(outputs, dim=1).squeeze(2), torch.stack(stops, dim=1).squeeze(-1)
+        res = res.masked_fill(pad_mask, PAD_IDX)
+        res_stop = res.masked_fill(pad_mask, PAD_IDX)
+        return res, res_stop
+
+    def decode_sequence(self, target, hidden_state, enc_output, enc_ctxt_mask, teacher_ratio=1):
+        """
+        For easier training!  target is teacher forcing [batch_size x seq_len x num_mels]
+        """
+        batch_size, max_out_len = target.shape[0], target.shape[1]
+        outputs = []
+        stops = []
+        # get a all 0 frame for first timestep
+        input_ = torch.zeros((batch_size, 1, target.shape[2]), device=target.device)
+        for i in range(max_out_len):
+            (dec_out, stop_pred), hidden_state = self.decode(input_, hidden_state, enc_output, enc_ctxt_mask)
+            stops.append(stop_pred)
+            outputs.append(dec_out)
+            if random.random() < teacher_ratio and i + 1 < max_out_len:
+                input_ = target[:, i+1, :].unsqueeze(1)
+            else:
+                input_ = outputs[-1]
+        return torch.stack(outputs, dim=1).squeeze(), torch.stack(stops, dim=1).squeeze()
+
+    def decode(self, input_, hidden_state, enc_output, enc_ctxt_mask):
+        """
+        NOTE: input_ should be [batch_size x 1 x num_mels] as this is autoregressive 
+            decoding (seq2seq).
+        Returns:
+            - dec_output: a tensor of shape [batch_size x 1 x num_mels]
+                + For text, dim will be #phonemes which is a distribution over 
+                    phonemes (needs to be log_softmaxed!)
+                + For speech, dim will be the ??? which is something log filter related
+            - dec_hidden: (h_n, c_n) from decoder LSTM
+        """
+        input_ = self.prenet(input_)
+        dec_output, dec_hidden = self.decoder(input_, hidden_state, enc_output, enc_ctxt_mask)
+        return self.postprocess(dec_output), dec_hidden
 
+    def postprocess(self, dec_output, distrib=False):
+        return self.postnet(dec_output)
 
-class TextTransformer(TextNet):
+    def forward(self, input_, mel_input):
+        encoder_outputs, latent_hidden, pad_mask = model.encode(input_)
+        pred, stop_pred = model.decode_sequence(mel_input, latent_hidden, encoder_outputs, pad_mask)
```

Comment:
Should this be `model` instead of `self`?

Reply:
Yeah it absolutely should be, I just copied and pasted this from the training code lol